compiler/rustc_hir/src/hir.rs

```diff
@@ -1532,6 +1532,64 @@ impl Expr<'_> {
         }
         expr
     }
+
+    pub fn can_have_side_effects(&self) -> bool {
+        match self.peel_drop_temps().kind {
+            ExprKind::Path(_) | ExprKind::Lit(_) => false,
+            ExprKind::Type(base, _)
+            | ExprKind::Unary(UnOp::UnDeref, base)
+            | ExprKind::Field(base, _)
+            | ExprKind::Index(base, _) => {
+                // This isn't exactly true for `Index`, but we are using this method exclusively
```

Comment:
nit: i personally think we can and should extend this to all `Unary` operations.
Isn't `Deref` also theoretically able to have sideeffects?

Reply:
It isn't but it is "_culturally_" side-effect free. If you saw code like `{ *foo; }` it would raise your eye brows, wouldn't it?

Reply:
yeah, the question was mostly why you specifically pointed out `Index` in the comment as the sime is also true for `Deref` and other unary operators